### PR TITLE
[12.0][MIG] onchange helper

### PR DESCRIPTION
**Pull request by sbejaoui**

### CODE REVIEW

File: onchange_helper/models/base.py
Comment (sbejaoui) [inline onchange_helper/models/base.py]: @lmignon , What do you think about this?

File: onchange_helper/models/base.py
Comment (sbejaoui) [inline onchange_helper/models/base.py]: @lmignon , What do you think about this?

File: onchange_helper/models/base.py
Comment (sbejaoui) [inline onchange_helper/models/base.py:45]: @lmignon , What do you think about this?

Comment (lmignon) [reply onchange_helper/models/base.py:45]: @sbejaoui  LGTM and will avoid 1 useless SQL query on each call to the method

File: onchange_helper/readme/CONTRIBUTORS.rst
Comment (lmignon) [inline onchange_helper/readme/CONTRIBUTORS.rst:3]: @sbejaoui You can add yourself into the contributors list

**Review by lmignon — APPROVED** [review]
Thank you @sbejaoui LGTM (Code review only)

File: onchange_helper/readme/CONTRIBUTORS.rst
Comment (sbejaoui) [reply onchange_helper/readme/CONTRIBUTORS.rst:3]: Done

**Review by sbejaoui — COMMENTED** [review]
@lmignon , I pushed some changes. can you check them please.

**Comment by lmignon** [comment]
@astirpe @sebastienbeau @florian-dacosta Could you take a look to the two last commits PLZ?
https://github.com/OCA/server-tools/pull/1461/commits/a22f10e2ad4df24d585284742491a2edcce9955a
https://github.com/OCA/server-tools/pull/1461/commits/0f4f8bbd74769a782710e66769d202702f3a99b5
These changes looks to me very useful and should be back ported into previous versions IMO.

File: onchange_helper/models/base.py
Comment (lmignon) [inline onchange_helper/models/base.py]: @sbejaoui You should take a look at https://github.com/OCA/server-tools/pull/1484/files

Comment (sbejaoui) [reply onchange_helper/models/base.py]: done, thanks :)

File: onchange_helper/models/base.py
Comment (lmignon) [inline onchange_helper/models/base.py:55]: @sbejaoui Do you remember the motivations behind the lines from 55 to 56? IMO we must exclude non stored fields. I observe that if I put all the fields of my model, my non stored fields are recomputed even if not accessed by an onchange...

Comment (sbejaoui) [reply onchange_helper/models/base.py:55]: We try to get values from existing record or defaults one to fill missing data in `all_values` passed to `play_onchange`.
https://github.com/OCA/server-tools/blob/176916f536793047ed7f9355f2ee56d9872735c9/onchange_helper/models/base.py#L43-L50
IMO, `field_name: self[field_name] ` trigger the compute process.

Comment (lmignon) [reply onchange_helper/models/base.py:55]: But do we really need all values.... We should let the ORM compute the one required once accessed. In my case, these lines trigger the compute of useless field and cause performance issue.

Comment (sbejaoui) [reply onchange_helper/models/base.py:55]: I think we can exclude computed fields.

Comment (lmignon) [reply onchange_helper/models/base.py:55]: I've the same feeling.

**Comment by bealdav** [comment]
Hi, thanks for your work.

I have a use case that doesn't work with this code.

**Float fields with a value of 0.0 is evaluated to False and then is not returned to vals**

It's is not easy to reproduce with unit tests on res.partner.

My case :

I import csv of purchase line with only specified product_id and qty

I want this onchange played called by onchange_product_id()
https://github.com/odoo/odoo/blob/339bf145ced6fb30f10d6d2941cfdaadf594d426/addons/purchase/models/purchase.py#L630

```python
vals = self.env['purchase.order.line'].play_onchanges(vals, ['product_id'])
```
`price_unit` is equal to 0.0, and don't reach vals but is mandatory for purchase lines.

I'll investigate deeper.

**Comment by lmignon** [comment]
@bealdav I'm refactoring this module in V10 to improve the performances... In the same time I'm adding a lot of tests. I've a corner case that doesn't pass the tests (recursive onchange on on2many relation). I'll also add a test for this UC to see if the new code is more robust. https://github.com/OCA/server-tools/pull/1510/files

**Comment by bealdav** [comment]
@lmignon thanks a lot

**Review by njeudy — APPROVED** [review]
Just install and test on a copy of a prod database.

Ok for me it works as needed !

**Review by ThomasBinsfeld — APPROVED** [review]

**Comment by rousseldenis** [comment]
@OCA/core-maintainers

**Pull request merged**
